backend/canisters/exchange_client/impl/src/exchanges/icdex.rs

```diff
@@ -0,0 +1,164 @@
+use crate::exchanges::Exchange;
+use async_trait::async_trait;
+use candid::{CandidType, Nat, Principal};
+use canister_client::make_c2c_call;
+use exchange_client_canister::{CancelOrderRequest, MakeOrderRequest, OrderType};
+use ic_cdk::api::call::{CallResult, RejectionCode};
+use ic_icrc1::endpoints::TransferArg;
+use ic_ledger_types::{AccountIdentifier, Subaccount};
+use serde::Serialize;
+use types::CanisterId;
+
+pub struct ICDexClient {
+    this_canister_id: CanisterId,
+    dex_canister_id: CanisterId,
+    icp_ledger_canister_id: CanisterId,
+    chat_ledger_canister_id: CanisterId,
+    unit_size: u64,
+}
+
+impl ICDexClient {
+    pub fn new(
+        this_canister_id: CanisterId,
+        dex_canister_id: CanisterId,
+        icp_ledger_canister_id: CanisterId,
+        chat_ledger_canister_id: CanisterId,
+        unit_size: u64,
+    ) -> ICDexClient {
+        ICDexClient {
+            this_canister_id,
+            dex_canister_id,
+            icp_ledger_canister_id,
+            chat_ledger_canister_id,
+            unit_size,
+        }
+    }
+
+    // TODO remove the `unwrap`s
+    pub async fn make_order(&self, caller: Principal, order: MakeOrderRequest) -> CallResult<()> {
+        let subaccount = subaccount(caller);
+        let from_account = AccountIdentifier::new(&self.this_canister_id, &Subaccount(subaccount));
+
+        let get_account_response: CallResult<(ic_icrc1::Account, String, Nat, Subaccount)> = make_c2c_call(
```

Comment:
@Hindsight said this could be generated - although he didn't say how. Worth chasing up again?

> TxAccount can be generated without calling the interface, it needs 3 parameters:
> 1. the principal of the caller
> 2. nonce
> 3. the canister principal of the trading pair
> Note that the nonce for each account is incremental and not random. If the nonce of a transaction is wrong, you need to get it again. getTxAccount() can query the latest nonce

Reply:
Yeah we would need to store the last used nonce and put that into the calculation. I think this is a follow on improvement we can make. Making trades is really low volume (currently) so I think its fine to go out using the simpler (but less performant) method.